repository_service_tuf/cli/artifact/download.py

```diff
@@ -0,0 +1,263 @@
+# SPDX-FileCopyrightText: 2023 VMware Inc
+#
+# SPDX-License-Identifier: MIT
+
+
+import base64
+import os
+import sys
+from hashlib import sha256
+from pathlib import Path
+from typing import Optional
+from urllib import request
+from urllib.parse import urlparse
+
+from click import Context
+from tuf.api.exceptions import DownloadError, RepositoryError
+from tuf.ngclient import Updater
+
+from repository_service_tuf.cli import click, console
+from repository_service_tuf.cli.artifact import artifact
+from repository_service_tuf.helpers.hash_bins import find_hash_bin
+
+
+def update_artifact_name_with_hash_prefix(artifact_name: str) -> str:
+    artifact_file_name = artifact_name.split("/")[-1]
+    hash_prefix = find_hash_bin(artifact_file_name)
+    new_artifact_file_name = hash_prefix + "-" + artifact_file_name
+    splitted = artifact_name.rsplit("/", 1)
+    if len(splitted) > 1:
+        return artifact_name.rsplit("/", 1)[0] + "/" + new_artifact_file_name
+    else:
+        # Annotating as it's covered, but codecov doesn't
+        # recognize that properly
+        return new_artifact_file_name  # pragma: no cover
+
+
+def decode_trusted_root(root) -> str:
+    root_encoded = bytes(root, "utf-8")
+    return base64.b64decode(root_encoded).decode("utf-8")
+
+
+def build_metadata_dir(metadata_url: str) -> str:
+    """build a unique and reproducible directory name for the repository url"""
+    name = sha256(metadata_url.encode()).hexdigest()[:8]
+    return f"{Path.home()}/.local/share/rstuf/{name}"
+
+
+def init_tofu(metadata_url: str, root_url: Optional[str]) -> bool:
+    """Initialize local trusted metadata (Trust-On-First-Use) and create a
+    directory for downloads"""
+    metadata_dir = build_metadata_dir(metadata_url)
+
+    os.makedirs(metadata_dir, exist_ok=True)
+
+    if not root_url:
+        root_url = f"{metadata_url}/1.root.json"
+    try:
+        parsed_root_url = urlparse(root_url)
+        if parsed_root_url.scheme and parsed_root_url.netloc:
+            request.urlretrieve(root_url, f"{metadata_dir}/root.json")  # nosec
+        else:
+            console.print(
+                f"Failed to parse {root_url}: ",
+                f"{parsed_root_url.scheme}, {parsed_root_url.netloc}",
+            )
+    except OSError:  # pragma: no cover
+        console.print(f"Failed to download initial root from {root_url}")
+        return False  # pragma: no cover
+
+    console.print(
+        f"Trust-on-First-Use: Initialized new root in {metadata_dir}"
+    )
+    return True
+
+
+def download_artifact(
+    metadata_url: Optional[str],
+    artifacts_url: Optional[str],
+    artifact_name: str,
+    root: Optional[str],
```

Comment:
`root` is still `Optional` when it shouldn't be.

Reply:
We can have an already existing `root.json` key, in which case we don't use the parameter, or we only use it for `init_tofu()` otherwise, where we check the parameter and retrieve a key if it's missing.